webapp/channels/src/components/advanced_text_editor/use_rewrite.tsx

```diff
@@ -0,0 +1,467 @@
+        setPrompt('');
+    }, []);
+
+    const undoMessage = useCallback(() => {
+        handleDraftChange({
+            ...draft,
+            message: originalMessage,
+        }, {instant: true});
+        focusTextbox();
+        resetState();
+    }, [draft, handleDraftChange, originalMessage, focusTextbox, resetState]);
+
+    const regenerateMessage = useCallback(() => {
+        setPrompt(lastPrompt);
+        handleDraftChange({
+            ...draft,
+            message: originalMessage,
+        }, {instant: true});
+        if (lastAction) {
+            handleRewrite(lastAction, lastAction === RewriteAction.CUSTOM ? lastPrompt : undefined);
+        }
+    }, [draft, handleRewrite, originalMessage, lastAction, lastPrompt, handleDraftChange]);
+
+    const cancelProcessing = useCallback(() => {
+        setIsProcessing(false);
+        setOriginalMessage('');
+        currentPromiseRef.current = undefined;
+    }, []);
+
+    const handleCustomPromptKeyDown = useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
+        if (e.key === ' ') {
+            e.stopPropagation();
+            return;
+        }
+        if (e.key === 'Enter') {
+            e.stopPropagation();
+            handleRewrite(RewriteAction.CUSTOM, prompt);
+        }
+    }, [handleRewrite, prompt]);
+
+    const handleMenuAction = useCallback((action: RewriteAction) => {
+        return () => handleRewrite(action);
+    }, [handleRewrite]);
+
+    useEffect(() => {
+        dispatch(getAgentsAction());
+    }, [dispatch]);
+
+    useEffect(() => {
+        if (agents && agents.length > 0 && !selectedAgentId) {
+            setSelectedAgentId(agents[0].id);
+        }
+    }, [agents, selectedAgentId]);
+
+    useEffect(() => {
+        if (isMenuOpen && !draft.message.trim()) {
+            customPromptRef.current?.focus();
+        }
+    }, [isMenuOpen, draft.message]);
+
+    // This adds an overlay to the textbox to
+    // indicate that the AI is rewriting the message.
+    // It might belong in the text editor - but since
+    // it's only used here, it's simpler to keep it here.
+    useEffect(() => {
+        if (!isProcessing || !textboxRef.current) {
+            return undefined;
+        }
+
+        const inputBox = textboxRef.current.getInputBox();
+        const wrapper = inputBox?.parentElement;
+        if (!wrapper) {
+            return undefined;
+        }
+
+        const overlay = document.createElement('div');
+        overlay.className = 'ai-rewrite-overlay';
+        wrapper.appendChild(overlay);
+
+        return () => {
+            if (wrapper.contains(overlay)) {
+                wrapper.removeChild(overlay);
+            }
+        };
+    }, [isProcessing, textboxRef]);
+
+    useEffect(() => {
+        if (isProcessing || draft.message.trim()) {
+            return;
+        }
+        resetState();
+    }, [draft.message, isProcessing, resetState]);
+
+    return {
+        additionalControl: useMemo(() => {
```

Comment:
Would it be worth splitting the component out of the hook? That way you can add some unit tests for the component in isolation